File: core/src/commonMain/kotlin/io/silencelen/andvari/core/crypto/SharedGrant.kt
package io.silencelen.andvari.core.crypto

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/** spec 01 §6 — the sealed shared-vault grant payload (canonical JSON: exact key order, no whitespace). */
@Serializable
data class SharedGrantPayload(
    @SerialName("v") val v: Int,
    @SerialName("vaultId") val vaultId: String,
    @SerialName("vk") val vk: String,
)

/**
 * Shared-vault key grants (spec 01 §6). A member grant seals the vault key to the
 * member's X25519 identity public key with `crypto_box_seal` (anonymous, no AD — the
 * vaultId rides inside the payload and the recipient re-checks it). The server stores
 * only the opaque `sealedVk` ciphertext; it can neither read nor redirect it. Sibling of
 * [Escrow]; the fingerprint construction is shared.
 */
object SharedGrant {
    private val json = Json { ignoreUnknownKeys = true }

    /** Canonical bytes — string template guarantees key order and no whitespace (matches web). */
    fun canonicalPayload(vaultId: String, vk: ByteArray): ByteArray =
        """{"v":1,"vaultId":"$vaultId","vk":"${Bytes.toB64(vk)}"}""".encodeToByteArray()

    fun seal(crypto: CryptoProvider, memberIdentityPub: ByteArray, vaultId: String, vk: ByteArray): ByteArray =
        crypto.sealTo(memberIdentityPub, canonicalPayload(vaultId, vk))

    /**
     * Open a sealed grant and return the 32-byte VK. The recipient MUST bind the sealed
     * payload to the grant row it arrived on: [expectedVaultId] must equal the payload's
     * vaultId (a malicious server cannot forge the seal, but this rejects a seal replayed
     * onto the wrong grant row).
     */
    fun open(
        crypto: CryptoProvider,
        memberIdentityPub: ByteArray,
        memberIdentityPriv: ByteArray,
        expectedVaultId: String,
        sealed: ByteArray,
    ): ByteArray {
        val payload = json.decodeFromString<SharedGrantPayload>(
            crypto.sealOpen(memberIdentityPub, memberIdentityPriv, sealed).decodeToString(),
        )
        if (payload.v != 1) throw CryptoException("unsupported shared-grant payload version ${payload.v}")
        if (payload.vaultId != expectedVaultId) throw CryptoException("shared grant vaultId mismatch")
        val vk = Bytes.fromB64(payload.vk)
        if (vk.size != 32) throw CryptoException("shared grant VK is not 32 bytes")
        return vk
    }

    /**
     * spec 01 §5 — identity fingerprint = lowercase hex SHA-256 of the identity public key.
     * CALLERS MUST pass the SEED-DERIVED pubkey (boxKeypairFromSeed(seed).publicKey), never
     * the server-supplied identityPub — else a malicious server can pass an out-of-band check.
     */
    fun fingerprint(crypto: CryptoProvider, identityPub: ByteArray): String =
        Bytes.toHexLower(crypto.sha256(identityPub))

    fun shortFingerprint(crypto: CryptoProvider, identityPub: ByteArray): String =
        fingerprint(crypto, identityPub).take(16)
}

File: web/src/crypto/sharedgrant.ts
import { fromB64, fromUtf8, toB64, toHexLower, utf8 } from "./bytes";
import { sealOpen, sealTo, sha256 } from "./provider";
import { CryptoError } from "./sodium";

/** spec 01 §6 — sealed shared-vault grant; mirrors core SharedGrant.kt byte-for-byte. */
export interface SharedGrantPayload {
  v: number;
  vaultId: string;
  vk: string;
}

/** Canonical bytes — string template guarantees key order and no whitespace (matches Kotlin). */
export function canonicalGrantPayload(vaultId: string, vk: Uint8Array): Uint8Array {
  return utf8(`{"v":1,"vaultId":"${vaultId}","vk":"${toB64(vk)}"}`);
}

export function sealSharedGrant(memberIdentityPub: Uint8Array, vaultId: string, vk: Uint8Array): Uint8Array {
  return sealTo(memberIdentityPub, canonicalGrantPayload(vaultId, vk));
}

/**
 * Open a sealed grant → the 32-byte VK. The payload's vaultId MUST equal the grant row's
 * vaultId (binds the seal to the row it arrived on).
 */
export function openSharedGrant(
  memberIdentityPub: Uint8Array,
  memberIdentityPriv: Uint8Array,
  expectedVaultId: string,
  sealed: Uint8Array,
): Uint8Array {
  const raw = sealOpen(memberIdentityPub, memberIdentityPriv, sealed);
  let payload: SharedGrantPayload;
  try {
    payload = JSON.parse(fromUtf8(raw)) as SharedGrantPayload;
  } catch (e) {
    throw new CryptoError("shared grant payload is not JSON", { cause: e });
  }
  if (payload.v !== 1) throw new CryptoError(`unsupported shared-grant payload version ${payload.v}`);
  if (payload.vaultId !== expectedVaultId) throw new CryptoError("shared grant vaultId mismatch");
  const vk = fromB64(payload.vk);
  if (vk.length !== 32) throw new CryptoError("shared grant VK is not 32 bytes");
  return vk;
}

/**
 * spec 01 §5 — identity fingerprint = lowercase hex SHA-256 of the identity public key.
 * MUST be computed over the SEED-DERIVED pubkey, never the server-supplied identityPub.
 */
export async function identityFingerprint(identityPub: Uint8Array): Promise<string> {
  return toHexLower(await sha256(identityPub));
}

export async function shortIdentityFingerprint(identityPub: Uint8Array): Promise<string> {
  return (await identityFingerprint(identityPub)).slice(0, 16);
}
